plugins/cfu/fu-cfu-module.c

```diff
@@ -0,0 +1,110 @@
+/*#
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-cfu-module.h"
+
+struct _FuCfuModule {
+	FuDevice parent_instance;
+	guint8 component_id;
+	guint8 bank;
+};
+
+G_DEFINE_TYPE(FuCfuModule, fu_cfu_module, FU_TYPE_DEVICE)
+
+static void
+fu_cfu_module_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuCfuModule *self = FU_CFU_MODULE(device);
+	fu_common_string_append_kx(str, idt, "Bank", self->bank);
+	fu_common_string_append_kx(str, idt, "ComponentId", self->component_id);
+}
+
+gboolean
+fu_cfu_module_setup(FuCfuModule *self, const guint8 *buf, gsize bufsz, gsize offset, GError **error)
+{
+	FuDevice *parent = fu_device_get_proxy(FU_DEVICE(self));
+	guint32 version_raw = 0;
+	guint8 tmp = 0;
+	g_autofree gchar *instance_id0 = NULL;
+	g_autofree gchar *instance_id1 = NULL;
+	g_autofree gchar *instance_id2 = NULL;
+	g_autofree gchar *logical_id = NULL;
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *version = NULL;
+
+	/* component ID */
+	if (!fu_common_read_uint8_safe(buf, bufsz, offset + 0x5, &self->component_id, error))
+		return FALSE;
+	name = g_strdup_printf("%s (0x%02X)", fu_device_get_name(parent), self->component_id);
+	fu_device_set_name(FU_DEVICE(self), name);
+
+	/* these GUIDs may cause the name or version-format to be overwritten */
+	instance_id0 = g_strdup_printf("USB\\VID_%04X&PID_%04X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)));
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id0);
+	instance_id1 = g_strdup_printf("USB\\VID_%04X&PID_%04X&CID_%02X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)),
+				       self->component_id);
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id1);
+
+	/* bank */
+	if (!fu_common_read_uint8_safe(buf, bufsz, offset + 0x4, &tmp, error))
+		return FALSE;
+	self->bank = tmp & 0b11;
+	instance_id2 = g_strdup_printf("USB\\VID_%04X&PID_%04X&CID_%02X&BANK_%01X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)),
+				       self->component_id,
+				       self->bank);
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id2);
+
+	/* version */
+	if (!fu_common_read_uint32_safe(buf, bufsz, offset, &version_raw, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	fu_device_set_version_raw(FU_DEVICE(self), version_raw);
+	version = fu_common_version_from_uint32(version_raw,
+						fu_device_get_version_format(FU_DEVICE(self)));
+
+	/* logical ID */
```

Comment:
the spec has callouts for some vendor specific field usages (https://docs.microsoft.com/en-us/windows-hardware/drivers/cfu/cfu-specification#5-cfu-protocol-packet-format).  We probably need to be thinking about how those fit into the architecture.  IE perhaps CFU should _not_ be it's own plugin, but a type of device that other plugins can inherit from or a library function, especially if they want to use some of those vendor specific usages in a specific way.

Reply:
> perhaps CFU should not be it's own plugin, but a type of device that other plugins can inherit from

That's a good idea, but maybe when we have at least *one* user of the CFU code -- unlike now where we don't even have that. Perhaps CFU could be implemented as a GObject interface rather than a concrete object, but without a consumer it feels like a ton of [exported] code. This plugin was meant more as a theoretical "how would we implement CFU" rather than something that I'm desperate to merge for new hardware support. Maybe making it derivable at this stage and adding the GtkDoc markup might be a good interim solution.